files/blockinfile.py

```diff
@@ -0,0 +1,292 @@
+    valid = not validate
+    if validate:
+        if "%s" not in validate:
+            module.fail_json(msg="validate must contain %%s: %s" % (validate))
+        (rc, out, err) = module.run_command(validate % tmpfile)
+        valid = rc == 0
+        if rc != 0:
+            module.fail_json(msg='failed to validate: '
+                                 'rc:%s error:%s' % (rc, err))
+    if valid:
+        module.atomic_move(tmpfile, dest)
+
+
+def check_file_attrs(module, changed, message):
+
+    file_args = module.load_file_common_arguments(module.params)
+    if module.set_file_attributes_if_different(file_args, False):
+
+        if changed:
+            message += " and "
+        changed = True
+        message += "ownership, perms or SE linux context changed"
+
+    return message, changed
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            dest=dict(required=True, aliases=['name', 'destfile']),
+            state=dict(default='present', choices=['absent', 'present']),
+            marker=dict(default='# {mark} ANSIBLE MANAGED BLOCK', type='str'),
+            block=dict(default='', type='str', aliases=['content']),
+            insertafter=dict(default=None),
+            insertbefore=dict(default=None),
+            create=dict(default=False, type='bool'),
+            backup=dict(default=False, type='bool'),
+            validate=dict(default=None, type='str'),
+        ),
+        mutually_exclusive=[['insertbefore', 'insertafter']],
+        add_file_common_args=True,
+        supports_check_mode=True
+    )
+
+    params = module.params
+    dest = os.path.expanduser(params['dest'])
+    if module.boolean(params.get('follow', None)):
+        dest = os.path.realpath(dest)
+
+    if os.path.isdir(dest):
+        module.fail_json(rc=256,
+                         msg='Destination %s is a directory !' % dest)
+
+    if not os.path.exists(dest):
+        if not module.boolean(params['create']):
+            module.fail_json(rc=257,
+                             msg='Destination %s does not exist !' % dest)
+        original = None
+        lines = []
+    else:
+        f = open(dest, 'rb')
+        original = f.read()
+        f.close()
+        lines = original.splitlines()
+
+    insertbefore = params['insertbefore']
+    insertafter = params['insertafter']
+    block = params['block']
+    marker = params['marker']
+    present = params['state'] == 'present'
+
+    if insertbefore is None and insertafter is None:
+        insertafter = 'EOF'
+
+    if insertafter not in (None, 'EOF'):
+        insertre = re.compile(insertafter)
+    elif insertbefore not in (None, 'BOF'):
+        insertre = re.compile(insertbefore)
+    else:
+        insertre = None
+
+    marker0 = re.sub(r'{mark}', 'BEGIN', marker)
+    marker1 = re.sub(r'{mark}', 'END', marker)
+    if present and block:
+        # Escape seqeuences like '\n' need to be handled in Ansible 1.x
+        if ANSIBLE_VERSION.startswith('1.'):
```

Comment:
So this could be removed after a bit (since in the official repository, it won't be shipped with ansible-1.x anyway?)